orttraining/orttraining/training_api/interfaces.h

```diff
@@ -90,30 +109,72 @@ class Module {
 
 // Internal state
 struct ParameterOptimizerState {
-  int64_t step_;
-  float learning_rate_;
   // Per param optimizer state. E.g. For Adam and param_0, this would contain
   // {“Moment_1_param_0”:<value>, …},
   // It should be noted that the names should only be maintained to correlate with
   // the graph inputs for the optimizer graph
-  std::map<std::string, OrtValue> states_;
+  std::unordered_map<std::string, std::shared_ptr<OrtValue>> states_;
 };
 
-struct OptimizerState {
+struct GroupOptimizerState {
   // overall state related to optimizer
   int64_t step_;
   float learning_rate_;
-  std::unordered_map<std::string, ParameterOptimizerState> optimizer_states_;
+  std::unordered_map<std::string, ParameterOptimizerState> param_named_optimizer_states_;
 };
 
+struct OptimizerCheckpointStates {
+ public:
+  std::unordered_map<std::string, std::shared_ptr<GroupOptimizerState>> group_named_optimizer_states;
+  const DataTransferManager* optimizer_session_data_transfer_mgr_;
+};
+
+static Status CreateOrtValueFromOrtValue(
+    const OrtValue& src_ort_value,
+    OrtValue& dest_ort_value,
+    onnxruntime::InferenceSession* sess) {
+  const Tensor& tensor = src_ort_value.Get<Tensor>();
+  AllocatorPtr allocator = sess->GetAllocator(tensor.Location());
+  const TensorShape& tensor_shape = tensor.Shape();
+  MLDataType element_type = tensor.DataType();
+
+  auto p_tensor = std::make_unique<Tensor>(element_type, tensor_shape, allocator);
+  dest_ort_value.Init(p_tensor.release(), DataTypeImpl::GetType<Tensor>(), DataTypeImpl::GetType<Tensor>()->GetDeleteFunc());
```

Comment:
Do you need to initialize this with some value

Reply:
Oh sorry just recalled, I intend to leave the implementation to you. :) If you already did that, I will just add a todo here.